compiler/typecheck/TcDaml.hs

```diff
@@ -13,34 +11,310 @@ import OccName
+extractDamlInfoFromGblEnv env =
+  foldMap extractDamlInfoFromTyThing (typeEnvElts $ tcg_type_env env) `mappend`
+    foldMap extractDamlInfoFromClsInst (instEnvElts $ tcg_inst_env env)
+
+extractDamlInfoFromEPS :: ExternalPackageState -> DamlInfo
+extractDamlInfoFromEPS eps =
+  foldMap extractDamlInfoFromTyThing (typeEnvElts $ eps_PTE eps) `mappend`
+    foldMap extractDamlInfoFromClsInst (instEnvElts $ eps_inst_env eps)
+
+extractDamlInfoFromTyThing :: TyThing -> DamlInfo
+extractDamlInfoFromTyThing tything =
+  case tything of
+    AnId id ->
+      mempty
+        { methods = mapMaybe matchMethod [id]
+        , implementations = mapMaybe matchImplements [id]
+        , choices = mapMaybe matchChoice [id]
+        }
+    ATyCon tycon ->
+      mempty
+        { templates = mapMaybe matchTemplate [tycon]
+        , interfaces = mapMaybe matchInterface [tycon]
+        }
+    _ -> mempty
+  where
+    matchTemplate, matchInterface :: TyCon -> Maybe Name
+    matchTemplate = tyconWithConstraint "DamlTemplate"
+    matchInterface = tyconWithConstraint "DamlInterface"
+
+    tyconWithConstraint :: String -> TyCon -> Maybe Name
+    tyconWithConstraint targetName tycon
+      | isAlgTyCon tycon
+      , let isMatchingLoneConstraint type_
+              | Just (loneConstraint, []) <- splitTyConApp_maybe type_
+              , similarName targetName (tyConName loneConstraint)
+              = True
+              | otherwise
+              = False
+      , any isMatchingLoneConstraint (tyConStupidTheta tycon)
+      = Just $ tyConName tycon
+      | otherwise
+      = Nothing
+
+    matchChoice :: Id -> Maybe (Name, Name)
+    matchChoice identifier
+      | TyConApp headTyCon [contractType, choiceType, returnType] <- varType identifier
+      , similarName "HasExercise" (tyConName headTyCon)
+      , Just (contractTyCon, []) <- splitTyConApp_maybe contractType
+      , Just (choiceTyCon, []) <- splitTyConApp_maybe choiceType
+      = Just (tyConName contractTyCon, tyConName choiceTyCon)
+      | otherwise
+      = Nothing
+
+    matchMethod :: Id -> Maybe (FastString, (Name, Type))
+    matchMethod identifier
+      | TyConApp headTyCon [contractType, methodNameType, returnType] <- varType identifier
+      , similarName "HasMethod" (tyConName headTyCon)
+      , LitTy (StrTyLit methodName) <- methodNameType
+      , Just (contractTyCon, []) <- splitTyConApp_maybe contractType
+      = Just (methodName, (tyConName contractTyCon, returnType))
+      | otherwise
+      = Nothing
+
+    matchImplements :: Id -> Maybe (Name, Name)
+    matchImplements identifier
+      | TyConApp headTyCon [templateType, interfaceType] <- varType identifier
+      , similarName "HasToInterface" (tyConName headTyCon)
+      , Just (templateTyCon, []) <- splitTyConApp_maybe templateType
+      , Just (interfaceTyCon, []) <- splitTyConApp_maybe interfaceType
+      = Just (tyConName templateTyCon, tyConName interfaceTyCon)
+      | otherwise
+      = Nothing
```

Comment:
It hasn't been necessary yet, but later we might want to know the parent of the `interface instance`; in that case we would match on 

```haskell
  | TyConApp headTyCon [parentType, interfaceType, templateType]
  , similarName "InterfaceInstance" (tyConName headTyCon)
  , {- ... -}
```

Reply:
Because this isn't an instance, I've been spending some time trying to get this to work by extracting the type signature from FunBind identifiers - it represents a large enough change that I'm pushing this to a future PR.